src/controllers/match2/index.ts

```diff
@@ -101,6 +94,76 @@ export class Match2Controller extends Controller {
 
     return responseWithAliases(match2)
   }
+
+  /**
+   * An optimiser creates the match2 {match2Id} on-chain. The match2 is now viewable to other members.
+   * @summary Create a new match2 on-chain
+   * @param match2Id The match2's identifier
+   */
+  @Post('{match2Id}/proposal')
+  @Response<NotFound>(404, 'Item not found')
+  @SuccessResponse('201')
+  public async createMatch2OnChain(@Path() match2Id: UUID): Promise<TransactionResponse> {
+    const [match2] = await this.db.getMatch2(match2Id)
+    if (!match2) throw new NotFound('match2')
+    if (match2.state !== Match2State.proposed) throw new BadRequest(`Match2 must have state: ${Match2State.proposed}`)
+
+    const [demandA] = await this.db.getDemand(match2.demandA)
+    validatePreOnChain(demandA, DemandSubtype.order, 'DemandA')
+
+    const [demandB] = await this.db.getDemand(match2.demandB)
+    validatePreOnChain(demandB, DemandSubtype.capacity, 'DemandB')
+
+    const [transaction] = await this.db.insertTransaction({
+      token_type: TokenType.MATCH2,
+      local_id: match2Id,
+      state: TransactionState.submitted,
+    })
+
+    // temp - until there is a blockchain watcher, need to await runProcess to know token IDs
+    const tokenIds = await runProcess(match2Propose(match2, demandA, demandB))
+    await this.db.updateTransaction(transaction.id, { state: TransactionState.finalised })
+
+    // match2-propose returns 3 token IDs
+    await observeTokenId(TokenType.DEMAND, match2.demandA, tokenIds[0], false) // order
+    await observeTokenId(TokenType.DEMAND, match2.demandB, tokenIds[1], false) // capacity
+    await observeTokenId(TokenType.MATCH2, match2.id, tokenIds[2], true) // match2
```

Comment:
three times calling the same function?

Reply:
it's temporary, pretending to be a blockchain watcher/indexer that's looping over tokens as they appear on chain